include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h

```diff
@@ -168,39 +319,112 @@ class __sort_global_kernel;
 template <typename... _Name>
 class __sort_copy_back_kernel;
 
+template <typename _IndexT>
+struct __parallel_sort_submitter_caller
+{
+    template <typename _ExecutionPolicy, typename _Range, typename _Compare, typename _LeafSorterT>
+    auto
+    operator()(_ExecutionPolicy&& __exec, _Range&& __rng, _Compare __comp, _LeafSorterT& __leaf_sorter)
+    {
+        using _CustomName = oneapi::dpl::__internal::__policy_kernel_name<_ExecutionPolicy>;
+        using _LeafDPWI = std::integral_constant<std::uint16_t, _LeafSorterT::__data_per_workitem>;
+        // TODO: split the submitter into multiple ones to avoid extra compilation of kernels
+        // - _LeafSortKernel does not need _IndexT
+        // - _GlobalSortKernel does not need _LeafDPWI
+        // - _CopyBackKernel does not need either of them
+        using _LeafSortKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+            __sort_leaf_kernel<_CustomName, _IndexT, _LeafDPWI>>;
+        using _GlobalSortKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+            __sort_global_kernel<_CustomName, _IndexT, _LeafDPWI>>;
+        using _CopyBackKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+            __sort_copy_back_kernel<_CustomName, _IndexT, _LeafDPWI>>;
+
+        return __parallel_sort_submitter<_IndexT, _LeafSortKernel, _GlobalSortKernel, _CopyBackKernel>()(
+            std::forward<_ExecutionPolicy>(__exec), std::forward<_Range>(__rng), __comp, __leaf_sorter);
+    }
+};
+
+template <typename _IndexT, typename _ExecutionPolicy, typename _Range, typename _Compare>
+auto
+__submit_selecting_leaf(_ExecutionPolicy&& __exec, _Range&& __rng, _Compare __comp)
+{
+    // 8 is the maximum reasonable value for bubble sub-group sorter due to algorithm complexity
+    // TODO: reconsider the value if another algorithm is used,
+    //       or an internal cap is set (e.g. sorting 2 sequences of 4/8 items each)
+    using _Leaf8 = __leaf_sorter<8, std::decay_t<_Range>, _Compare>;
+    using _Leaf4 = __leaf_sorter<4, std::decay_t<_Range>, _Compare>;
+    // 2 is the smallest reasonable value for merge-path group sorter since it loads 2 values at least
+    using _Leaf2 = __leaf_sorter<2, std::decay_t<_Range>, _Compare>;
+
+    using _Tp = oneapi::dpl::__internal::__value_t<_Range>;
+
+    const std::size_t __n = __rng.size();
+    auto __device = __exec.queue().get_device();
+
+    const std::size_t __max_wg_size = __device.template get_info<sycl::info::device::max_work_group_size>();
+    std::size_t __wg_size = oneapi::dpl::__internal::__dpl_bit_floor(__max_wg_size);
+
+    const bool __is_cpu = __device.is_cpu();
+    std::uint32_t __max_sg_size{};
+    if (__is_cpu)
+    {
+        const auto __sg_sizes = __device.template get_info<sycl::info::device::sub_group_sizes>();
+        __max_sg_size = __sg_sizes.empty() ? 1 : *std::max_element(__sg_sizes.begin(), __sg_sizes.end());
```

Comment:
This is for a case if the call returns an empty vector. The specification says nothing whether it can be empty or not. 
However, implicitly, it seems it should not be empty, since the minimum number of sub-groups in a work-group is 1:

> info::device::max_num_sub_groups: Returns the maximum number of sub-groups in a work-group for any kernel executed on the device. The minimum value is 1.

I will get rid of this check.

Reply:
Done